### PR TITLE
DEP: add a `[typing]` extra dependency target for type-check-only dependencies

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Following @nstarman's suggestion in https://github.com/astropy/astropy/pull/15686#issuecomment-1843690039

ping @hamogu

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -66,8 +66,12 @@ recommended = [
     "scipy>=1.5",
     "matplotlib>=3.3,!=3.4.0,!=3.5.2",
 ]
+type = [
```
Comment (eerovaher) [inline pyproject.toml:69]:
```suggestion
typing = [
```

**Review by eerovaher — COMMENTED** [review]
It's quite sensible to have a separate set of dependencies for typing so that only people who choose to install `astropy[type]` or `astropy[all]` will have the extra packages installed, but I'd prefer `astropy[typing]` over `astropy[type]`.

**Comment by neutrinoceros** [comment]
> I'd prefer astropy[typing] over astropy[type]

Me too ! what do you think @nstarman ?

**Comment by neutrinoceros** [comment]
done !

**Review by eerovaher — APPROVED** [review]

**Comment by nstarman** [comment]
Failure appears unrelated. Let's get this in!

**Review by nstarman — APPROVED** [review]

**Pull request merged**